diff --git a/src/com/cooliris/picasa/GDataClient.java b/src/com/cooliris/picasa/GDataClient.java
index a2e2be2..b156347 100644
--- a/src/com/cooliris/picasa/GDataClient.java
+++ b/src/com/cooliris/picasa/GDataClient.java
@@ -32,6 +32,7 @@
 import org.apache.http.conn.scheme.PlainSocketFactory;
 import org.apache.http.conn.scheme.Scheme;
 import org.apache.http.conn.scheme.SchemeRegistry;
+import org.apache.http.conn.ssl.SSLSocketFactory;
 import org.apache.http.entity.ByteArrayEntity;
 import org.apache.http.entity.InputStreamEntity;
 import org.apache.http.impl.client.DefaultHttpClient;
@@ -70,6 +71,7 @@
         // Register HTTP protocol.
         SchemeRegistry schemeRegistry = new SchemeRegistry();
         schemeRegistry.register(new Scheme("http", PlainSocketFactory.getSocketFactory(), 80));
+        schemeRegistry.register(new Scheme("https", SSLSocketFactory.getSocketFactory(), 443));

         // Create the connection manager.
         HTTP_CONNECTION_MANAGER = new ThreadSafeClientConnManager(params, schemeRegistry);